### PR TITLE
EC2 template updates

### PR DESCRIPTION
**Pull request by lamadome**
Changes to ec2 quickstart and module to enble features described in prd https://docs.google.com/document/d/1zleMpseFKMok_dZ7VVfjLqKwVUPiJT12p-gWvTlEbSA/edit

### CODE REVIEW

File: hcp-ui-templates/ec2-existing-vpc/main.tf [hunk 1]
```diff
@@ -60,19 +62,38 @@ resource "hcp_consul_cluster" "main" {
 resource "hcp_consul_cluster_root_token" "token" {
   cluster_id = hcp_consul_cluster.main.id
 }
+resource "tls_private_key" "ssh" {
+  algorithm = "RSA"
+  rsa_bits  = 4096
+}
+rresource "aws_key_pair" "hcp_ec2_key_pair" {
```
Comment (jjtimmons) [inline hcp-ui-templates/ec2-existing-vpc/main.tf:69]:
```suggestion

resource "aws_key_pair" "hcp_ec2_key_pair" {
```

File: modules/hcp-ec2-client/main.tf [hunk 2]
```diff
@@ -3,7 +3,7 @@ data "aws_ami" "ubuntu" {
 
   filter {
     name   = "name"
-    values = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-*"]
+    values = ["ubuntu/images/hvm-ssd/ubuntu-hirsute-21.04-amd64-server-*"]
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/main.tf:6]: we'll want to leave this unchanged. My understanding is that Focal has some very long-term support, [where other images don't.](https://wiki.ubuntu.com/Releases) We just had a mini-incident where another image version wasn't supported https://github.com/hashicorp/terraform-aws-hcp-consul/pull/48

File: modules/hcp-ec2-client/main.tf [hunk 3]
```diff
@@ -44,18 +48,49 @@ resource "aws_security_group_rule" "allow_http_inbound" {
...
+}
+
+#The bellow is to setup the instance profile and iam role to enable SSM
+resource "aws_iam_instance_profile" "hcp-ec2-iam-profile" {
+  role        = aws_iam_role.hcp-ec2-iam-role.name
+  name_prefix = "hcp_ec2_profile"
+}
+
+resource "aws_iam_role" "hcp-ec2-iam-role" {
+  #name        = "hcp-ec2-ssm-role"
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/main.tf:61]: is this name still needed?

File: modules/hcp-ec2-client/main.tf [hunk 4]
```diff
@@ -44,18 +48,49 @@ resource "aws_security_group_rule" "allow_http_inbound" {
   protocol    = "tcp"
   cidr_blocks = var.allowed_http_cidr_blocks
 
-  security_group_id = var.security_group_id
+  security_group_id = aws_security_group.hcp_consul_ec2.id
+}
+
+#The bellow is to setup the instance profile and iam role to enable SSM
+resource "aws_iam_instance_profile" "hcp-ec2-iam-profile" {
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/main.tf:55]: nit: use all underscore for resource names, and no need to respecify `iam_profile`: https://www.terraform-best-practices.com/naming

```suggestion
resource "aws_iam_instance_profile" "hcp_ec2" {
```

File: hcp-ui-templates/ec2-existing-vpc/main.tf [hunk 5]
```diff
@@ -94,7 +115,34 @@ output "nomad_url" {
 output "hashicups_url" {
   value = "http://${module.aws_ec2_consul_client.public_ip}"
 }
-
+output "consul_export" {
+  value     = <<EOF
+  export CONSUL_HTTP_ADDR="${hcp_consul_cluster.main.consul_public_endpoint_url}"
+  export CONSUL_HTTP_TOKEN="${hcp_consul_cluster_root_token.token.secret_id}"
```
Comment (jjtimmons) [inline hcp-ui-templates/ec2-existing-vpc/main.tf:121]: I think it'd be easier to export each of these separately for cleanliness. So like:

```terraform
output "consul_http_addr" {
...
}

output "consul_http_token" {
...
}
```

The `export $VAR=` is a shell-specific command that isn't really relevant to terraform, and we can combine that with the `export` bit in your `how_to_connect` output below

File: hcp-ui-templates/ec2-existing-vpc/main.tf [hunk 6]
```diff
@@ -94,7 +115,34 @@ output "nomad_url" {
...
 output "next_steps" {
-  value = "Hashicups Application will be ready in ~2 minutes. Use 'terraform output consul_root_token' to retrieve the root token."
+  value = local.install_demo_app ? "Hashicups Application will be ready in ~2 minutes. Use 'terraform output consul_root_token' to retrieve the root token." : null
+}
+output "ssh_to_client" {
+  value = local.ssh ? "ssh -i ${local_file.ssh_key.filename} ubuntu@${module.aws_ec2_consul_client.public_ip}" : null
+}
+output "connect_with_ssm" {
+  value = "aws ssm start-session --target ${module.aws_ec2_consul_client.host_id} --region ${local.vpc_region}"
+}
```
Comment (jjtimmons) [inline hcp-ui-templates/ec2-existing-vpc/main.tf:140]: I can see an argument either way, but why not keep these in the `how_to_connect` output below? (as new lines)

With one output snippet per connection method, we'll have a bunch of output strings. It seems like we could include each of these snippets in the how to connect section and reduce the clutter of the number of exported outputs

File: modules/hcp-ec2-client/main.tf [hunk 7]
```diff
@@ -44,18 +48,49 @@ resource "aws_security_group_rule" "allow_http_inbound" {
...
+}
+  EOF
+  name_prefix        = "hcp_ec2_role"
+}
+
+resource "aws_iam_role_policy_attachment" "dev-resources-ssm-policy" {
+  role       = aws_iam_role.hcp-ec2-iam-role.name
+  policy_arn = "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore"
+}
+
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/main.tf:80]: nit: extra new line here

File: modules/hcp-ec2-client/nomad-job.tf [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+  create_duration = "15s"
+}
+resource "nomad_job" "hashicups-frontend-v2" {
+  count   = var.install_demo_app ? 1 : 0
+  jobspec = file("${path.module}/templates/hashicups-frontend-v2.nomad")
+  hcl2 {
+    enabled = true
+  }
+  depends_on = [
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/nomad-job.tf:40]: we install both?

File: hcp-ui-templates/ec2-existing-vpc/main.tf [hunk 9]
```diff
@@ -60,19 +62,38 @@ resource "hcp_consul_cluster" "main" {
 resource "hcp_consul_cluster_root_token" "token" {
   cluster_id = hcp_consul_cluster.main.id
 }
+resource "tls_private_key" "ssh" {
+  algorithm = "RSA"
```
[the remark this answers is not in the document]
Comment (lamadome) [reply hcp-ui-templates/ec2-existing-vpc/main.tf:66]: yes but in this case i just save it in the directory where the TF is ran. So i dont think wed run into this issue. Thanks for mentioning!

File: modules/hcp-ec2-client/main.tf (hunk 2 above)
Comment (lamadome) [reply modules/hcp-ec2-client/main.tf:6]: Yeh i remember that, this change was unwanted on my part.  will revert.

File: modules/hcp-ec2-client/nomad-job.tf (hunk 8 above)
Comment (lamadome) [reply modules/hcp-ec2-client/nomad-job.tf:40]: The reason i do that is that we rely on this for the learn tutorial on blue green and canary. I dont want to have to tell people how to deploy a second version set the tags etc... The goal here is just to show users the functionality

Comment (lamadome) [reply modules/hcp-ec2-client/nomad-job.tf:45]: Discussed during our meeting earlier

File: hcp-ui-templates/ec2-existing-vpc/main.tf (hunk 6 above)
Comment (lamadome) [reply hcp-ui-templates/ec2-existing-vpc/main.tf]: I've just tested this out and it is indeed nicer, however it is a big output snippet which i guess is fine. I'll keep this open so we can compare both once i've commit that change

File: examples/hcp-ec2-demo/output.tf [hunk 10]
```diff
@@ -20,5 +20,23 @@ output "hashicups_url" {
...
+  ${var.install_demo_app ? "The demo app, HashiCups, is installed on a Nomad server we have deployed for your." : ""}
+  ${var.install_demo_app ? "To access Nomad using your local client run the following command" : ""}
+  ${var.install_demo_app ? "export NOMAD_HTTP_AUTH=nomad:$(terraform output consul_root_token)" : ""}
+  ${var.install_demo_app ? "export NOMAD_ADDR=http://${module.aws_ec2_consul_client.public_ip}:8081" : ""}
+
+  To access Consul from your local client run:
+  export CONSUL_HTTP_ADDR="${hcp_consul_cluster.main.consul_public_endpoint_url}"
+  export CONSUL_HTTP_TOKEN=$(terraform output consul_root_token)
+  
+  To connect to the ec2 instance deployed, you have 2 options:
```
Comment (jjtimmons) [inline examples/hcp-ec2-demo/output.tf:38]:
```suggestion
  To connect to the ec2 instance deployed: 
```

nit: but the 2nd option won't be available in future if not deploying the app

File: examples/hcp-ec2-demo/output.tf [hunk 11]
```diff
@@ -20,5 +20,23 @@ output "hashicups_url" {
 }
 
 output "next_steps" {
-  value = "Hashicups Application will be ready in ~2 minutes. Use 'terraform output consul_root_token' to retrieve the root token."
+  value = local.install_demo_app ? "Hashicups Application will be ready in ~2 minutes. Use 'terraform output consul_root_token' to retrieve the root token." : null
+}
+
+output "howto_connect" {
+  value = <<EOF
+  "In order to get access to both nomad and consul from the command line run the following commands:
```
Comment (jjtimmons) [inline examples/hcp-ec2-demo/output.tf:28]:
```suggestion
  To access Nomad and Consul from the command line run the following commands:
```

I don't think we'll need the starting double quote

File: examples/hcp-ec2-demo/variables.tf [hunk 12]
```diff
@@ -39,3 +39,16 @@ variable "tier" {
...
 }
+
+variable "ssh" {
+  type        = string
+  description = "Enable or disable SSH access via locally created certificate"
+  default     = true
+}
+
+variable "install_demo_app" {
+  type        = string
```
Comment (jjtimmons) [inline examples/hcp-ec2-demo/variables.tf:50]:
```suggestion
  type        = bool
```

File: examples/hcp-ec2-demo/variables.tf [hunk 13]
```diff
@@ -39,3 +39,16 @@ variable "tier" {
   description = "The HCP Consul tier to use when creating a Consul cluster"
   default     = "development"
 }
+
+variable "ssh" {
+  type        = string
```
Comment (jjtimmons) [inline examples/hcp-ec2-demo/variables.tf:44]:
```suggestion
  type        = bool
```

File: modules/hcp-ec2-client/variables.tf [hunk 14]
```diff
@@ -8,7 +8,10 @@ variable "allowed_ssh_cidr_blocks" {
   type        = list(string)
   default     = []
 }
-
+variable "ssh_keyname" {
+  description = "key pair name for ssh connection"
+  default     = ""
+}
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/variables.tf:14]:
```suggestion

variable "ssh_keyname" {
  type        = string
  description = "key pair name for ssh connection"
  default     = ""
}

```

File: modules/hcp-ec2-client/templates/hashicups.nomad [hunk 15]
```diff
@@ -25,35 +29,90 @@ variable "product_db_port" {
...
+                proxy_pass http://public_api_upstream;
+              }
+            
+          }
+          EOF
+          destination = "local/default.conf"
       }
     }
   }
-
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/templates/hashicups.nomad:56]: nit: can we leave all these newlines between the nomad stanzas? Imo they help a bit with readability

File: modules/hcp-ec2-client/templates/hashicups.nomad [hunk 16]
```diff
@@ -3,6 +3,10 @@ variable "frontend_port" {
   default     = 3000
 }
 
+variable "nginx_port" {
+  type        = number
+  default     = 80
+}
 variable "public_api_port" {
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/templates/hashicups.nomad:10]:
```suggestion
}

variable "public_api_port" {
```

File: modules/hcp-ec2-client/nomad-job.tf [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  jobspec  = file("${path.module}/templates/hashicups.nomad")
+  hcl2 {
+    enabled = true
+  }
+  depends_on = [
+    time_sleep.wait_for_client
+  ]
+}
+
+resource "nomad_job" "hashicups-frontend" {
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/nomad-job.tf:26]:
```suggestion
resource "nomad_job" "hashicups_frontend" {
```

File: modules/hcp-ec2-client/nomad-job.tf [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+resource "time_sleep" "wait_15_seconds" {
+
+  depends_on = [nomad_job.hashicups-frontend]
+
+  create_duration = "15s"
+}
+
+resource "nomad_job" "hashicups-frontend-v2" {
```
Comment (jjtimmons) [inline modules/hcp-ec2-client/nomad-job.tf:42]:
```suggestion
resource "nomad_job" "hashicups_frontend_v2" {
```

File: examples/hcp-ec2-demo/output.tf (hunk 11 above)
Comment (jjtimmons) [reply examples/hcp-ec2-demo/output.tf:28]: Can we just delete this line? I don't think `Nomad` will be there if `!var.install_demo_app` and in that case it's just Consul which already has some documentation below

**Review by jjtimmons — COMMENTED** [review]
mostly just nits/clean-up sort of things, but otherwise LGTM

File: examples/hcp-ec2-demo/output.tf (hunk 10 above)
Comment (lamadome) [reply examples/hcp-ec2-demo/output.tf:38]: But you will always want to connect to the ec2 instance right? On ec2 if we do not install the app we want to give the user a consul client(ec2) ready to go and registered with consul.

File: examples/hcp-ec2-demo/output.tf (hunk 11 above)
Comment (lamadome) [reply examples/hcp-ec2-demo/output.tf:28]: Good point, deleting the line

File: examples/hcp-ec2-demo/variables.tf (hunk 13 above)
Comment (lamadome) [reply examples/hcp-ec2-demo/variables.tf:44]: copy paste error 👎

**Review by jjtimmons — APPROVED** [review]

**Pull request merged**